### PR TITLE
log the Firecracker version at start-up

### PR DESCRIPTION
**Pull request by serban300**
Issue # , if available: #400 

Description of changes: log the Firecracker version at start-up

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by serban300** [comment]
Thank you for taking a look on the changes. Here are my answers:

1. Passing the version at `preinit` and saving it is the easiest option. The alternative would be to put the firecracker version in a context (like `shared_info`) and making sure it reaches `vmm/src/lib.rs -> init_logger` where it will be passed to `LOGGER.init()`. Would this approach be better ? 

1. I agree that it can be a bit confusing. The `instance_id` that is passed at `preinit` is temporary, and in Firecracker it is actually needed, so I can't remove it. But I could make it `Option` . I can also clarify this in the documentation.

**Comment by serban300** [comment]
I addressed the comments.

File: logger/src/lib.rs [hunk 1]
```diff
@@ -9,27 +9,27 @@
 //! # Enabling logging
 //! The first step in making use of the logging functionality, is to explicitly initialize it. Any
 //! intent to log either human readable content or metrics will silently fail until `LOGGER.init()`
-//! is called and it returns `Ok`. The logging subsystem is considered to be initialized when both
-//! the log and metric destinations have been configured to be a pipe. In other words,
-//! `LOGGER.init(<ID>, None, None)` can be called any number of times, up until the first call with
-//! both parameters set to `Some`. Any call to the `LOGGER.init()` following that will fail with an
+//! is called and it returns `Ok`. Any call to the `LOGGER.init()` following that will fail with an
```
Comment (aghecenco) [inline logger/src/lib.rs]: Please document `preinit` as well.

File: logger/src/lib.rs [hunk 2]
```diff
@@ -9,27 +9,27 @@
...
 //! extern crate logger;
-//! use logger::LOGGER;
+//! use logger::{AppInfo, LOGGER};
 //! use std::ops::Deref;
 //!
 //! fn main() {
-//!     // Initialize the logger. If there is no path to a FIFO provided, the `LOGGER` logs the
-//!     // human readable content to stdout and stderr depending on the log level.
-//!     if let Err(e) = LOGGER.deref().init("MY-INSTANCE", None, None, vec![]) {
+//!     // Optionally preinitialize the logger
```
Comment (aghecenco) [inline logger/src/lib.rs]: nit: add full stop to preserve comment style consistency

File: logger/src/lib.rs [hunk 3]
```diff
@@ -9,27 +9,27 @@
...
+//!     if let Err(e) = LOGGER.deref().preinit(Some("MY-INSTANCE".to_string())) {
 //!         println!("Could not initialize the log subsystem: {:?}", e);
 //!         return;
 //!     }
+//!     if let Err(e) = LOGGER.deref().init(&AppInfo::new("Firecracker", "1.0"), "MY-INSTANCE", "/tmp/log".to_string(), "/tmp/metrics".to_string(), vec![]) {
+//!            println!("Could not initialize the log subsystem: {:?}", e);
```
Comment (aghecenco) [inline logger/src/lib.rs]:
```suggestion
//!            println!("Could not initialize the log subsystem: {}", e);
```

File: logger/src/lib.rs [hunk 4]
```diff
@@ -62,7 +62,8 @@
 //!          libc::mkfifo(metrics.as_bytes().as_ptr() as *const i8, 0o644);
 //!     }
 //!     // Initialize the logger to log to a FIFO that was created beforehand.
-//!     assert!(LOGGER.deref().init("MY-INSTANCE", Some(logs), Some(metrics), vec![]).is_ok());
+//!     assert!(LOGGER.deref().preinit(Some("MY-INSTANCE".to_string())).is_ok());
```
Comment (aghecenco) [inline logger/src/lib.rs]: nit: add a comment about `preinit` as well.

File: logger/src/lib.rs [hunk 5]
```diff
@@ -221,6 +223,26 @@ impl FromStr for LogOption {
     }
 }
 
+/// A structure containing info about the App that uses the logger
```
Comment (aghecenco) [inline logger/src/lib.rs]: nit: add full stop

File: logger/src/lib.rs [hunk 6]
```diff
@@ -221,6 +223,26 @@ impl FromStr for LogOption {
...
+pub struct AppInfo {
+    name: String,
+    version: String,
+}
+
+impl AppInfo {
+    /// Creates a new instance of AppInfo
```
Comment (aghecenco) [inline logger/src/lib.rs]: nit: add full stop

File: logger/src/lib.rs [hunk 7]
```diff
@@ -502,6 +524,66 @@ impl Logger {
         Ok(())
     }
 
+    /// Performs the most basic steps in order to enable the logger to write to StdOut or StdErr
```
Comment (aghecenco) [inline logger/src/lib.rs]: Please keep the first line of the comment as a short summary, as for `init`. For instance, `Preconfigure the logger prior to initialization.`

File: logger/src/lib.rs [hunk 8]
```diff
@@ -511,37 +593,48 @@ impl Logger {
...
+            }
+            INITIALIZED => {
+                // If the logger was already initialized, an error will be returned.
+                METRICS.logger.log_fails.inc();
+                return Err(LoggerError::AlreadyInitialized);
+            }
+            _ => {}
         }
 
+        // operate on the RwLock within a block in order to avoid poisoning the lock
```
Comment (aghecenco) [inline logger/src/lib.rs]: nit: format the comment to match the style of other comments in this function

File: tests/integration_tests/functional/test_logging.py [hunk 9]
```diff
@@ -47,6 +47,9 @@ def check_log_message(log_str, instance_id, level, show_level, show_origin):
     e.g.:
     `2018-09-09T12:52:00.123456789 [MYID:WARN:/path/to/file.rs:52] warning`
     """
+    if log_str.startswith("Running Firecracker"):
+        return
```
Comment (aghecenco) [inline tests/integration_tests/functional/test_logging.py]: Why stop the test here?

File: vmm/src/vmm_config/instance_info.rs [hunk 10]
```diff
@@ -38,6 +38,8 @@ pub struct InstanceInfo {
     pub id: String,
     /// The state of the microVM.
     pub state: InstanceState,
+    /// The version of the Firecracker container that runs the microVM
+    pub firecracker_version: String,
```
Comment (aghecenco) [inline vmm/src/vmm_config/instance_info.rs]: Can you rename `firecracker_version` to something not Firecracker-specific?

Comment (serban300) [reply vmm/src/vmm_config/instance_info.rs]: I renamed it to vmm_version. I don't know if it's the best name.

**Comment by aghecenco** [comment]
Can you also add an entry in CHANGELOG.md?

**Comment by dianpopa** [comment]
> I personally prefer passing the version to `init` directly and not storing it in the logger. I'll let @dianpopa chime in with her opinion as well 😃

I actually liked a lot the idea of having an `init` and a `preinit`. The reason is that it eliminates the need for the additional logic that needed to be there in order to check whether or not the FIFOs are an `Option` or not. I think it makes the code easier to read because it leaves no doubt as to what will be the destination of the logs before the user configures logging and what is the destination after the user configures it (i.e `preinit` does not need them so it will go to stdout, `init` needs them both to be successful).

**Comment by aghecenco** [comment]
> > I personally prefer passing the version to `init` directly and not storing it in the logger. I'll let @dianpopa chime in with her opinion as well smiley
> 
> I actually liked a lot the idea of having an `init` and a `preinit`. The reason is that it eliminates the need for the additional logic that needed to be there in order to check whether or not the FIFOs are an `Option` or not. I think it makes the code easier to read because it leaves no doubt as to what will be the destination of the logs before the user configures logging and what is the destination after the user configures it (i.e `preinit` does not need them so it will go to stdout, `init` needs them both to be successful).

What about storing the version in the logger object? It's only used in `init`, so why keep it around instead of supplying it as a parameter of `init`, logging it there and letting it go?

**Comment by dianpopa** [comment]
> > > I personally prefer passing the version to `init` directly and not storing it in the logger. I'll let @dianpopa chime in with her opinion as well smiley
> > 
> > 
> > I actually liked a lot the idea of having an `init` and a `preinit`. The reason is that it eliminates the need for the additional logic that needed to be there in order to check whether or not the FIFOs are an `Option` or not. I think it makes the code easier to read because it leaves no doubt as to what will be the destination of the logs before the user configures logging and what is the destination after the user configures it (i.e `preinit` does not need them so it will go to stdout, `init` needs them both to be successful).
> 
> What about storing the version in the logger object? It's only used in `init`, so why keep it around instead of supplying it as a parameter of `init`, logging it there and letting it go?

I think I do not understand your question about storing the version in the logger object. As far as I see in the code the init function has a parameter `AppInfo` where the version is specified (plus the name of the app, in our case Firecracker). As far as I understand, The `AppInfo` structure is only a way of being able to provide multiple parameters related to the app that uses the logger and I like the approach. This `AppInfo` gets created upon `init_logger` from the `vmm/src/lib.rs`. Similar to the `instance_id`, the `firecracker_version` gets retrieved from the `shared_info` created upon starting the vmm tthread and then passed to the `LOGGER.init(...)` function as a parameter. The version is logged with the `log_helper` which basically flushes things to the log FIFO provided. I do not see `AppInfo` being kept around anywhere.
Am I missing something? Where exactly is the version kept around?

**Comment by aghecenco** [comment]
> > > > I personally prefer passing the version to `init` directly and not storing it in the logger. I'll let @dianpopa chime in with her opinion as well smiley
> > > 
> > > 
> > > I actually liked a lot the idea of having an `init` and a `preinit`. The reason is that it eliminates the need for the additional logic that needed to be there in order to check whether or not the FIFOs are an `Option` or not. I think it makes the code easier to read because it leaves no doubt as to what will be the destination of the logs before the user configures logging and what is the destination after the user configures it (i.e `preinit` does not need them so it will go to stdout, `init` needs them both to be successful).
> > 
> > 
> > What about storing the version in the logger object? It's only used in `init`, so why keep it around instead of supplying it as a parameter of `init`, logging it there and letting it go?
> 
> I think I do not understand your question about storing the version in the logger object. As far as I see in the code the init function has a parameter `AppInfo` where the version is specified (plus the name of the app, in our case Firecracker). As far as I understand, The `AppInfo` structure is only a way of being able to provide multiple parameters related to the app that uses the logger and I like the approach. This `AppInfo` gets created upon `init_logger` from the `vmm/src/lib.rs`. Similar to the `instance_id`, the `firecracker_version` gets retrieved from the `shared_info` created upon starting the vmm tthread and then passed to the `LOGGER.init(...)` function as a parameter. The version is logged with the `log_helper` which basically flushes things to the log FIFO provided. I do not see `AppInfo` being kept around anywhere.
> Am I missing something? Where exactly is the version kept around?

Nope, you're not missing anything, I was. I still had in mind the previous version of the code, which kept stored the version in the logger config. I have no issue with `AppInfo` and it's true that separating `init` and `preinit` clarify the requirements.

File: logger/src/lib.rs [hunk 11]
```diff
@@ -502,46 +525,117 @@ impl Logger {
...
-            METRICS.logger.log_fails.inc();
-            return Err(LoggerError::AlreadyInitialized);
-        }
-
-        if (log_pipe.is_none() && metrics_pipe.is_some())
-            || (log_pipe.is_some() && metrics_pipe.is_none())
-        {
-            return Err(LoggerError::DifferentDestinations);
+        match STATE.compare_and_swap(UNINITIALIZED, INITIALIZING, Ordering::SeqCst) {
+            PREINITIALIZING => {
```
[the remark this answers is not in the document]
Comment (serban300) [reply logger/src/lib.rs]: Good ideea. I didn't notice this.

**Comment by dianpopa** [comment]
@serban300 you should use `git commit --amend` followed by `git push -f` instead of creating a new commit for each new change to the initial commit.

**Comment by serban300** [comment]
Oh sorry, I didn't realize that. I will try it next time.

**Pull request merged**